### ISSUE TITLE
Running Jetset on additional binaries

### ISSUE THREAD

**Issue by DNA-Broker** [aerosec/jetset#4]
Hi,
I am trying to run JetSet on an additional binary, but I am having no luck. I am trying to run it on the GPS tracker binary located at:

https://github.com/fuzzware-fuzzer/fuzzware-experiments/blob/main/02-comparison-with-state-of-the-art/uEmu/uEmu.GPSTracker/uEmu.GPSTracker.elf

I have set up a gps_config.py file with the following contents:

```
from device.base_device import Simmio
import angr
import cle

peripherals = Simmio( lower=0x40000000,
                 upper=0x5FFFFFFF,
                 name='peripherals')

regions = [peripherals]


def get_project():        
    p = angr.Project("../jetset_public_data/uEmu.GPSTracker.elf", auto_load_libs=False)
    return p

target = 0x0008480f
arch = "arm"
arch_num = 1
```
I have also added a gps entry to socs.py as well as a run_gps.sh script with the following:

```
#!/usr/bin/env bash
$( dirname "${BASH_SOURCE[0]}" )/build/arm-softmmu/qemu-system-arm -machine stm32f4 \
  -m 256 \
  -initial_pc 0x837a1 \
  -initial_sp 0x20088000 \
  -kernel $(dirname "${BASH_SOURCE[0]}")/../jetset_public_data/uEmu.GPSTracker.elf \
  -serial stdio \
  -d nochain -singlestep -S -qmp tcp:localhost:$1,server,nowait
```

In addition I have added the following to the makefile:
```
run_gps:
	source jetset_env/bin/activate && cd $(ENGINE_BASE) && python $(ENGINE_SCRIPT) --soc=gps -o=gps_device.c --cmdfile ../jetset_qemu/run_gps.sh --useFunctionPrologues
````

However, when I run it with `make run_gps` all I get is the follwing in terminal. 

![image](https://user-images.githubusercontent.com/118234302/201811671-c298eaaa-7037-459a-84be-2a92f88f145f.png)

This just runs forever, but xterm and qemu both open.

Is there something I am missing in order to get this running on additional binaries?

**Comment by maxwell-bland** [comment]
Hello, thanks for your effort.

One of three things is happening. These are research questions that I wanted to address in the original work but I've been sidetracked breaking redactions, haha. 

(1) There are function calls called "TCG helper operations" that also need to be taint-tracked inside QEMU for the symbolic execution to work correctly for some microarch (assembly) instructions and features (e.g. task-switching). No existing work has implemented taint tracking for all of these operations as the work required would be excessive (re-implementing an entire microarch), even some questionable _claims_ to having implemented SE support for QEMU automatically. You may have such an operation on your codepath and it may be important. You can instrument the TCG interpreter in QEMU to find out (see the other issue I opened on this repo).

(2) The search strategy is getting itself stuck into a bad position. In general, finding the correct program path n a binary is hard, but for whatever reason the academic security community doesn't like acknowledging this and I was ... discouraged .... to put it lightly, from being able to address this in the paper. One solution is to record the set of decisions that get made and walk along with them while performing some reverse engineering to figure out where it gets stuck, and implementing a different strategy.

For 1 and 2, I may also be able to get you access to a new project that I think solves these problems (and others), but is a radically different architecture and approach to the problem of firmware rehosting (and you may not want this, since it isn't "published" yet and ... given what I know of the academic process, may never be). Email me and I can send you the code.

(3) There's some bug in the taint tracking. During the process, I argued we should be making unit tests and was made fun of by the coauthors. To track it down, uncomment many of the print statements I made in the TCG interpreter and track down where either (1) a symbolic decision is not being made when it should be or (2) the wrong decision is being made, likely because of some incorrect symbolic constraints.

When/if you do find the solution, please make a pull request! I've wanted to work on Jetset but have not been able to for innumerable reasons.

Maxwell

**Comment by JohnKIMMI** [comment]
I ran the same firmware and encountered the same issue.

Although I haven't solved it yet, I discovered that the firmware's main function contains a delay(1) function.

While analyzing the process, I found that execution gets stuck at a wfi (Wait For Interrupt) instruction. The delay function is supposed to wait for 1ms and then proceed, but the timer does not seem to function properly during execution.

I suspect that the symbolic execution might not be handling the timer correctly.

**Comment by maxwell-bland** [comment]
Hi @JohnKIMMI ,thanks for the insight! It has been a while since I have looked at the code for this.

Ideally, we'd have the QEMU translator intercept WFI instructions during decoding, and then split the symbolic state within angr into X separate states for each interrupt vector location, each beginning at the entry point of the appropriate IVT, with the stack and symbolic taint tracking for variables on the stack propagated. IIRC, when an interrupt occurs, some set of GP registers, e.g. x0-xN are pushed to the stack.

So, I'd say you should just use some sort of quick hack here. I'd just manually tag that PC, and then have an override which triggers the magic necessary interrupt in the IVT (OR honestly, maybe just patch the binary).  I recall Evan wrote something like "we iterate through interrupts at random points in the code". The WFI stall should then be handled, but there may be a bug in this code w.r.t. this case. Overall, should not be too hard to fix.

tl;dr let me know if you need further pointers or this is enough, apologies I do not have more time/funding to maintain this project at this time.
